lib/path/default.nix

```
@@ -92,8 +106,82 @@ let
     # An empty string is not a valid relative path, so we need to return a `.` when we have no components
     (if components == [] then "." else concatStringsSep "/" components);
 
+
+  # Takes a Nix path value and deconstructs it into the filesystem root
+  # (generally `/`) and a subpath
+  deconstructPath =
+    let
+      go = components: path:
+        # If the parent of a path is the path itself, then it's a filesystem root
+        if path == dirOf path then { root = path; inherit components; }
+        else go ([ (baseNameOf path) ] ++ components) (dirOf path);
+    in go [];
+
+
+
 in /* No rec! Add dependencies on this file at the top. */ {
 
+  difference = paths:
+    let
+      # Deconstruct every path into its root + subpath
+      deconstructed = mapAttrsToList (name: value:
+        # Check each item to be an actual path
+        assert assertMsg (isPath value) "lib.path.difference: Given attribute ${name} is of type ${typeOf value}, but a path value was expected";
+        deconstructPath value // { inherit name value; }
+      ) paths;
+
+      first =
+        assert assertMsg
+          (paths != {})
+          "lib.path.difference: An empty attribute set was given, but was expecting a non-empty attribute set";
+        head deconstructed;
+
+      # The common root to all paths, errors if there are different roots
+      commonRoot =
+        # Fast happy path in case all roots are the same
+        if all (dec: dec.root == first.root) deconstructed then first.root
+        # Slower sad path when that's not the case and we need to throw an error
+        else let
+          trigger = foldl'
+            (skip: dec:
+              if dec.root == first.root then skip
+              else throw "lib.path.difference: Path ${first.name} = ${toString first.value} (root ${toString first.root}) has a different filesystem root than path ${toString dec.name} = ${toString dec.value} (root ${toString dec.root})")
+            null
+            deconstructed;
+          fallbackAbort =
+            abort "lib.path.difference: This should never happen and indicates a bug! Some paths don't have the same filesystem root but it couldn't be found!";
+        in seq trigger fallbackAbort;
+
+      goCommonAncestorLength = level:
+        let headComponent = elemAt first.components level; in
```

```suggestion
        let firstSegment = elemAt firstPath.segments index; in
```

I had a hard time grokking this function. This should make clearer what it's about. Paths don't have levels, it's a sequence, therefore items in the sequence have an index.